pytorch_lightning/loops/base.py

```diff
@@ -0,0 +1,59 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional
+from weakref import proxy
+
+import pytorch_lightning as pl
+
+
+class Loop(ABC):
+
+    def __init__(self):
+        self.iteration_count: int = 0
+        self.trainer: Optional['pl.Trainer'] = None
+
+    def connect(self, trainer, *args, **kwargs) -> None:
+        """Connects Loop with all the necessary things like connectors and accelerators"""
+        self.trainer = proxy(trainer)
+
+    @property
+    @abstractmethod
+    def done(self) -> bool:
+        """Property indicating when loop is finished"""
+
+    @abstractmethod
+    def reset(self) -> None:
+        pass
+
+    def run(self, *args: Any, **kwargs: Any) -> Any:
+        self.reset()
+        self.on_run_start(*args, **kwargs)
+
+        while not self.done:
+            try:
+                self.on_advance_start(*args, **kwargs)
+                self.advance(*args, **kwargs)
+                self.on_advance_end()
+                self.iteration_count = self.increment_iteration(self.iteration_count)
+            except StopIteration:
+                break
+
+        return self.on_run_end()
+
+    def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    def on_advance_start(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    @abstractmethod
+    def advance(self, *args: Any, **kwargs: Any) -> None:
+        """What to do within a single step"""
+
+    def on_advance_end(self) -> None:
+        pass
+
+    def on_run_end(self) -> Any:
```

Comment:
should we distinguish between loop teardown vs returning results? or do you think these will always happen together?

Reply:
once you leave the loop (execute `on_run_end`) you will *always* return results and teardown if necessary. There is no way to continue a run for now (you could skip some iterations based on the internal state/checkpoint, but not really continue a run once you left it).

Coming back means to start a new run. So IMO having them separate does not make sense (well we have to kinds of teardown, one for the run which should be performed in `on_run_end` and one of the persistent state (which should be performed upon object deletion/will be performed automatically by gc)

Reply:
> There is no way to continue a run for now

But we want to have a way in the future. Should we design it now to accommodate the separation between `teardown` and `end`?

Reply:
we could, so you want this?

```python
# in run:
...
results = self.on_run_end()
self.teardown()
return results
```

Reply:
Do you plan to resolve this here @awaelchli ? Add an empty teardown?

Reply:
added an empty teardown to the interface as shown above. 
Can be refined when resolving todo in #7938